### PR TITLE
Test, benchmark and optimize the input system

### PR DESCRIPTION
**Pull request by nitya-dev**
Adds full test coverage to key and input related systems. The tests are generic and might be missing some edge cases but those should be added as they pop up.

Adds benchmarks to prevent performance degradation with future changes. Again generic tests but they did point out some issues that are now resolved.

Tweaked the key manager to be faster in almost every use case:

- Removed a lot of branching behavior
- Removed the use of 'defer' due to the high cost

---

Over all speed up from original code is 3x on all benchmarks below latest result on my system:

```
 nitya > ⋯ > src > engo.io > engo > go test -bench=. -benchmem
PASS
BenchmarkInputMgr_AxisCleanState-4       2000000               760 ns/op               0 B/op          0 allocs/op
BenchmarkInputMgr_AxisFilledState-4      2000000               896 ns/op               0 B/op          0 allocs/op
BenchmarkInputMgr_ButtonCleanState-4     1000000              1071 ns/op               0 B/op          0 allocs/op
BenchmarkInputMgr_ButtonFilledState-4    1000000              1217 ns/op               0 B/op          0 allocs/op
BenchmarkKeyMgr_CleanState-4             5000000               327 ns/op               0 B/op          0 allocs/op
BenchmarkKeyMgr_FilledState-4            5000000               395 ns/op               0 B/op          0 allocs/op
BenchmarkKeyMgr_CleanUpdate-4            5000000               371 ns/op               0 B/op          0 allocs/op
BenchmarkKeyMgr_FilledUpdate-4           3000000               426 ns/op               0 B/op          0 allocs/op
BenchmarkKeyMgr_CleanSubOptimal-4        1000000              1104 ns/op               0 B/op          0 allocs/op
BenchmarkKeyMgr_FilledSubOptimal-4       1000000              1308 ns/op               0 B/op          0 allocs/op
ok      engo.io/engo    18.008s
```

No allocs means no garbage generated and that matters as well ;)

---

Edit: More changed and with tests in place this is now up for merge again. Going any further would need changes that belong in a separated commit.

### CODE REVIEW

**Comment by nitya-dev** [comment]
Maybe hold off on merge. I'm reworking the whole input mechanics for my own project, might want to merge that in staid if it is ready. Based of the current system and my imput manager in unity.

Key Changes:
* Key and mouse combo actions
* Mouse buttons behave like keys
* Events when actions change state
* Store state for registered actions once

'Actions' would be axis, button and sequence objects that are registered. Stored state would update all registered actions once so no complex methods need to be called when checking the state of an action inside update functions.

**Comment by nitya-dev** [comment]
Updated with test and benchmark. Initial changes gave +-10% speed increase, the last commit just see ...

Before:
```
$ nitya > ⋯ > src > engo.io > engo > go test -bench=.
PASS
BenchmarkKeyMgrCleanState-4              1000000              1102 ns/op
BenchmarkKeyMgrCleanUpdate-4             1000000              1149 ns/op
BenchmarkKeyMgrFilledState-4             1000000              1287 ns/op
BenchmarkKeyMgrFilledUpdate-4             300000              4082 ns/op
BenchmarkKeyMgrCleanStateSubOptimal-4     300000              4201 ns/op
ok      engo.io/engo    6.156s
```

After:
```
$ nitya > ⋯ > src > engo.io > engo > go test -bench=.
PASS
BenchmarkKeyMgrCleanState-4              5000000               318 ns/op
BenchmarkKeyMgrCleanUpdate-4             5000000               362 ns/op
BenchmarkKeyMgrFilledState-4             5000000               379 ns/op
BenchmarkKeyMgrFilledUpdate-4             500000              3129 ns/op
BenchmarkKeyMgrCleanStateSubOptimal-4    1000000              1121 ns/op
ok      engo.io/engo    9.114s
```

Feel free to post your results and thoughts, always interested to see numbers from other systems.

**Comment by otraore** [comment]
Thank you @nitya-dev . I didn't realize about the cost of defer until now (https://github.com/nitya-dev/engo/commit/6da5bb9a10a7cfd5a76ca294c67fc49c171c41ee#diff-3afb33cd4bbb018c7856457d229b4586L41 and talks with @hydroflame ).

File: axis_test.go [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+type axState struct {
+	value float32
+}
+
+type axKeyCfg struct {
+	Name  string
+	Pairs []AxisKeyPair
+}
+
+////////////////
```
Comment (paked) [inline axis_test.go]: This comment here is not idiomatic godoc. It should be removed

Comment (paked) [inline axis_test.go]: Same with all the other similar ones :D

**Comment by paked** [comment]
@nitya-dev Once you remove the non idiomatic comments, I'll be happy to merge this in!

File: .runtests.sh [hunk 2]
```diff
@@ -8,6 +8,9 @@ go get -t -v ./... || exit 1
 echo "Testing engo.io/engo using coveralls"
 $HOME/gopath/bin/goveralls -service=travis-ci
 
+echo "Testing engo.io/engo using 'go test'"
```
Comment (otraore) [inline .runtests.sh]: Change to "Testing engo.io/engo using 'go test -v -bench'" or something of that nature (to mark a difference between regular tests and benchmarks.

Comment (paked) [inline .runtests.sh]: What about "Testing and benchmarking engo.io/engo"

Comment (otraore) [inline .runtests.sh]: That sounds much better, let's go with that @paked

**Comment by otraore** [comment]
LGTM and I think it LGT@paked as well. Thank you for your contributions, merging it in!

**Comment by paked** [comment]
Thanks for the awesome PR @nitya-dev.
